docs/design-docs/01-oci-support.md

```diff
@@ -0,0 +1,338 @@
+kpt pkg get oci://us-docker.pkg.dev/the-project-id/the-repo-name/the-package//simple/example:v3 my-package/simple/my-example
+```
+
+### `kpt pkg update`
+
+The command for update is not changed, but when the `upstream` is `oci` then the `@VERSION` is used to change the `upstream` image's `tag` or `digest` value.
+
+`kpt pkg update @v14` and `kpt pkg update DIR@v14` will assign `:v14` as the new upstream image tag.
+
+`kpt pkg update @sha256:xxxxx` and `kpt pkg update DIR@sha256:xxxxx` will assign `@sha256:xxxxx` as the new upstream image digest.
+
+`kpt pkg update` and `kpt pkg update DIR` use the currently defined `upstream` image tag or digest unchanged.
+
+At that point, if the `upstream` is an `image:tag` that is to discover the current `image:digest` for tag, otherwise the `upstream` value for `image:digest` is used. In either case, the `upstreamLock` is changed to point at that new `image:digest`. 
+
+The package contents of the old and new `upstreamLock` image digest are fetched to temp folders, and are the basis of the 3-way merge to update the target package.
+
+### `kpt pkg diff`
+
+The `kpt pkg diff` command is identical to `kpt pkg update` in the way that `[PKG_PATH@VERSION]` argument is mapped to OCI concepts.
+
+### Command additions
+
+### ` kpt pkg push`
+
+Summary TBD
+
+### ` kpt pkg pull`
+
+Summary TBD
+
+## User Guide
+
+### Creating a package respository
+
+Before kpt packages can be pushed and pulled as OCI images, a suitable repository 
+must be created. Google Artifact Registry and Google Container Registry are both
+excellent choices.
+
+```shell
+# Choose names and locations
+LOCATION="us"
+PROJECT_ID="kpt-demo-73823"
+REPOSITORY_NAME="blueprints"
+
+# Base name for any images in this repository
+REPOSITORY="${LOCATION}-docker.pkg.dev/${PROJECT_ID}/${REPOSITORY_NAME}"
+
+# Create the repository
+gcloud artifacts repositories create --location="${LOCATION}" --repository-format=docker --project="${PROJECT_ID}" "${REPOSITORY_NAME}"
+```
+
+### Creating and pushing a new package
+
+Creating a new package is no different. But when ready to publish, the `kpt pkg push` command 
+is used instead of source control operations.
+
+```shell
+# A package in a new directory
+mkdir hello-world
+kpt pkg init hello-world --description="A simple blueprint"
+
+# Store the contents in the repository, tagged as v1
+kpt pkg push hello-world --image=${REPOSITORY}/hello-world:v1
+
+# The local files are not needed any more, pushing has stored them all
+rm -r hello-world
+```
+
+### Pulling and updating a package
+
+Because the package folder was discarded earlier, the `kpt pkg pull` command 
+is used to place the contents of a particular version at a location. These set of
+commands may be run in `Cloud Build` steps as well, if you are automating the
+publication of packages as part of a CI/CD process.
+
+```shell
+# Recreate the folder and extract the pulled image
+kpt pkg pull hello-world --image=${REPOSITORY}/hello-world:v1
```

Comment:
They are very similar, really, with key difference being `pull` wouldn't change Kptfile upstream values to point at the source.

In other words you can pull/push back and forth any number of times without any change to the actual package contents.

Because `oci` isn't using a cli like `git` to fetch and commit content - a different mechanism has to be available.

Reply:
Got it. So it is similar to `git clone` and `git push`. But I feel it is confusing to the user. What are the existing alternatives to pull and push OCI images? Can we punt these 2 commands for now(so don't add any commands in initial pass) and only augment the behavior of existing package commands to support OCI images ?

Reply:
There's some context in
https://github.com/GoogleContainerTools/kpt/issues/2557

Reply:
In retrospect, it would have been better to call "kpt pkg get" something else, like clone or fork.

Reply:
I suppose we could also add a flag to get that would fetch the package without changing the upstream info. That seems asymmetric with pull, though.

Reply:
I could also imagine a kpt pkg delete command.

Reply:
Thinking through:
Users should use `kpt pkg get` when they want to fork a package with the intention of keeping it up-to-date with the upstream. The user story is: I am a platform developer at company X and I want to take the upstream pkg say (nginx) and customize it for my company and publish it to my company registry (new target). So it will look like:

```
kpt pkg get registry.nginx/nginx:v1.0 companyx-nginx
cd companyx-nginx
modify loop
kpt pkg push registry.companyx/nginx:v1.0 .
```

For all rest use-cases, `kpt pkg pull` is sufficient.

> In retrospect, it would have been better to call "kpt pkg get" something else, like clone or fork.

I agree. Actually `get` feels closer to `read-only` copy (more like fetch immutable pkg).

Reply:
I realized we also update few more things to the pkg in `get`:

- change the name of the pkg to the destination directory
- Add `merge-key` directive to keep track of identity to ensure `pkg update` is able to do resource-merge
- upstream-info (and lock)

so in `pull` I am expecting only `lock` info to be updated.

Reply:
Yes, that flow looks right to me. An initial `get`-`alter`-`push` followed by `pull`-`update`-`push` for each change.

For the second case - in the `pull`-`update`-`push` - I'd expect the `upstreamLock` to be updated by the `update` and not the `pull`, right? Because I may be using `pull`-`anything`-`push` to make change as trivial as updating comment text and I wouldn't want other side effects.

Reply:
> For the second case - in the pull-update-push - I'd expect the upstreamLock to be updated by the update and not the pull, right?

Yes, you are right. updating my comment.

Reply:
I agree with the use-case for pull and push. Currently in the context of git, platform developers use git clone and git push instead of `kpt pkg pull` and `kpt pkg push` in the workflow above. So having this commands also makes sense in the context of git based workflows. Users would not need to use other tools to achieve the task. 

One open question I have here is does `kpt pkg push` push the upstream and upstreamLock section also to  `registry.companyx/nginx:v1.0` in the above example ? The best practice is that a consumer at any level should only know about its immediate upstream. If we delete that information during `kpt pkg push` how would day N scenario work for a platform developer in company X in the above workflow.

Also `kpt pkg copy` seems more intuitive as users will be confused with `kpt pkg get` and `kpt pkg pull` commands.

Reply:
Resolving this thread - get/pull difference has been clarified.